### PR TITLE
feat(decompress): Support Zstandard archive

### PR DESCRIPTION
**Pull request by worldyun**
scoop\bin> .\scoop.ps1 install openssh  

Installing 'zstd' (1.5.0) [64bit]
Loading zstd-v1.5.0-win64.zip from cache   
Checking hash of zstd-v1.5.0-win64.zip ... ok.
Extracting zstd-v1.5.0-win64.zip ... done.
Linking D:\scoop\apps\zstd\current => D:\scoop\apps\zstd\1.5.0
Creating shim for 'zstd'.
'zstd' (1.5.0) was installed successfully!
Installing 'openssh' (8.2p1-1) [64bit]
Loading bash-4.4.023-2-x86_64.pkg.tar.xz from cache   
Checking hash of bash-4.4.023-2-x86_64.pkg.tar.xz ... ok.
Loading gcc-libs-9.3.0-1-x86_64.pkg.tar.xz from cache   
Checking hash of gcc-libs-9.3.0-1-x86_64.pkg.tar.xz ... ok.
Loading heimdal-7.7.0-2-x86_64.pkg.tar.zst from cache      
Checking hash of heimdal-7.7.0-2-x86_64.pkg.tar.zst ... ok.
Loading heimdal-libs-7.7.0-2-x86_64.pkg.tar.zst from cache 
Checking hash of heimdal-libs-7.7.0-2-x86_64.pkg.tar.zst ... ok.
Loading libcrypt-2.1-2-x86_64.pkg.tar.xz from cache
Checking hash of libcrypt-2.1-2-x86_64.pkg.tar.xz ... ok.
Loading libdb-5.3.28-3-x86_64.pkg.tar.zst from cache     
Checking hash of libdb-5.3.28-3-x86_64.pkg.tar.zst ... ok. 
Loading libedit-20191231_3.1-1-x86_64.pkg.tar.xz from cache
Checking hash of libedit-20191231_3.1-1-x86_64.pkg.tar.xz ... ok.
Loading libopenssl-1.1.1.e-1-x86_64.pkg.tar.xz from cache        
Checking hash of libopenssl-1.1.1.e-1-x86_64.pkg.tar.xz ... ok.  
Loading libsqlite-3.30.0-1-x86_64.pkg.tar.xz from cache        
Checking hash of libsqlite-3.30.0-1-x86_64.pkg.tar.xz ... ok.
Loading msys2-runtime-3.1.4-1-x86_64.pkg.tar.zst from cache  
Checking hash of msys2-runtime-3.1.4-1-x86_64.pkg.tar.zst ... ok.
Loading ncurses-6.2-1-x86_64.pkg.tar.xz from cache
Checking hash of ncurses-6.2-1-x86_64.pkg.tar.xz ... ok.
Loading openssh-8.2p1-1-x86_64.pkg.tar.xz from cache    
Checking hash of openssh-8.2p1-1-x86_64.pkg.tar.xz ... ok.
Loading openssl-1.1.1.e-1-x86_64.pkg.tar.xz from cache    
Checking hash of openssl-1.1.1.e-1-x86_64.pkg.tar.xz ... ok.
Loading zlib-1.2.11-1-x86_64.pkg.tar.xz from cache      
Checking hash of zlib-1.2.11-1-x86_64.pkg.tar.xz ... ok.
Extracting bash-4.4.023-2-x86_64.pkg.tar.xz ... done.
Extracting gcc-libs-9.3.0-1-x86_64.pkg.tar.xz ... done.
Extracting heimdal-7.7.0-2-x86_64.pkg.tar.zst ... done.
Extracting heimdal-libs-7.7.0-2-x86_64.pkg.tar.zst ... done.
Extracting libcrypt-2.1-2-x86_64.pkg.tar.xz ... done.
Extracting libdb-5.3.28-3-x86_64.pkg.tar.zst ... done.
Extracting libedit-20191231_3.1-1-x86_64.pkg.tar.xz ... done.
Extracting libopenssl-1.1.1.e-1-x86_64.pkg.tar.xz ... done.
Extracting libsqlite-3.30.0-1-x86_64.pkg.tar.xz ... done.
Extracting msys2-runtime-3.1.4-1-x86_64.pkg.tar.zst ... done.
Extracting ncurses-6.2-1-x86_64.pkg.tar.xz ... done.
Extracting openssh-8.2p1-1-x86_64.pkg.tar.xz ... done.
Extracting openssl-1.1.1.e-1-x86_64.pkg.tar.xz ... done.
Extracting zlib-1.2.11-1-x86_64.pkg.tar.xz ... done.
Running pre-install script...
Junction created for D:\scoop\apps\openssh\8.2p1-1\home\wordy\.ssh <<===>> C:\Users\wordy\.ssh
Linking D:\scoop\apps\openssh\current => D:\scoop\apps\openssh\8.2p1-1
Creating shim for 'findssl.sh'.
Creating shim for 'scp'.
Creating shim for 'sftp'.
Creating shim for 'ssh'.
Creating shim for 'ssh-add'.
Creating shim for 'ssh-agent'.
Creating shim for 'ssh-copy-id'.
Creating shim for 'ssh-keygen'.
Creating shim for 'ssh-keyscan'.
Creating shim for 'sshd'.
'openssh' (8.2p1-1) was installed successfully!

### CODE REVIEW

**Comment by Valinor** [comment]
Question : 

When I reviewed (only to be informed), I see that : 

```
    if(!(Test-Path "$env:SCOOP_HELPERS\lessmsi\lessmsi.exe")) {
        Start-FileDownload 'https://github.com/activescott/lessmsi/releases/download/v1.6.3/lessmsi-v1.6.3.zip' -FileName "$env:SCOOP_HELPERS\lessmsi.zip"
        & 7z.exe x "$env:SCOOP_HELPERS\lessmsi.zip" -o"$env:SCOOP_HELPERS\lessmsi" -y
    }
    if(!(Test-Path "$env:SCOOP_HELPERS\innounp\innounp.exe")) {
        Start-FileDownload 'https://raw.githubusercontent.com/ScoopInstaller/Binary/master/innounp/innounp048.rar' -FileName "$env:SCOOP_HELPERS\innounp.rar"
        & 7z.exe x "$env:SCOOP_HELPERS\innounp.rar" -o"$env:SCOOP_HELPERS\innounp" -y
    }
```

**v1.6.3** for lessmsi and **innounp048.rar**. This is old (lessmsi is from 23 Nov 2018, innounp048 is from 2019-01-27).
We can currently found inounp50 and lessmsi 1.10. It is for test only, and I doesn't read all the code, maybe some upgrade are done after that.

**Review by rasa — APPROVED** [review]
Untested, but LGTM. I see no harm with merging it into the develop branch.

File: PSScriptAnalyzerSettings.psd1 [hunk 1]
```diff
@@ -5,7 +5,7 @@
     # Only diagnostic records of the specified severity will be generated.
     # Uncomment the following line if you only want Errors and Warnings but
     # not Information diagnostic records.
-    Severity = @('Error','Warning')
+    Severity = @('Error')
```
Comment (goostleek) [inline PSScriptAnalyzerSettings.psd1:8]: Could you elaborate a bit why we are decreasing the log level to `Error` only? I think it has a real value to have `Warning` as well. WDYT?

File: lib/install.ps1 [hunk 2]
```diff
@@ -598,6 +598,8 @@ function dl_urls($app, $version, $manifest, $bucket, $architecture, $dir, $use_c
             }
         } elseif(Test-7zipRequirement -File $fname) { # 7zip
             $extract_fn = 'Expand-7zipArchive'
+        }elseif(Test-ZstdRequirement -File $fname) { # Zstd
```
Comment (goostleek) [inline lib/install.ps1:601]:
```suggestion
        } elseif(Test-ZstdRequirement -File $fname) { # Zstd
```

File: test/bin/init.ps1 [hunk 3]
```diff
@@ -22,6 +22,10 @@ if ($env:CI_WINDOWS -eq $true) {
         Start-FileDownload 'https://raw.githubusercontent.com/ScoopInstaller/Binary/master/innounp/innounp048.rar' -FileName "$env:SCOOP_HELPERS\innounp.rar"
         & 7z.exe x "$env:SCOOP_HELPERS\innounp.rar" -o"$env:SCOOP_HELPERS\innounp" -y
     }
+    if(!(Test-Path "$env:SCOOP_HELPERS\zstd\zstd.exe")) {
```
Comment (goostleek) [inline test/bin/init.ps1:25]:
```suggestion
    if (!(Test-Path "$env:SCOOP_HELPERS\zstd\zstd.exe")) {
```

File: test/Scoop-Decompress.Tests.ps1 [hunk 4]
```diff
@@ -75,6 +75,32 @@ Describe 'Decompression function' -Tag 'Scoop', 'Decompress' {
         }
     }
 
+    # Context "zstd extraction" {
```
Comment (goostleek) [inline test/Scoop-Decompress.Tests.ps1:78]: Why we have these tests commented out?

**Comment by niheaven** [comment]
I've just rewrite `Expand-ZstdArchive` and make some minor fix in other files.

Works with `diffutils` and will merge into develop, please test the code with other manifest.

**Comment by niheaven** [comment]
@goostleek The whitespaces could be fixed maybe in another PR that format all the code base.

File: lib/decompress.ps1 [hunk 5]
```diff
@@ -104,6 +122,45 @@ function Expand-7zipArchive {
...
+        $DestinationPath = (Split-Path $Path),
+        [String]
+        $ExtractDir,
+        [Parameter(ValueFromRemainingArguments = $true)]
+        [String]
+        $Switches,
+        [ValidateSet("All")]
+        [String]
+        $Overwrite,
+        [Switch]
```
[the remark this answers is not in the document]
Comment (rasa) [reply lib/decompress.ps1:141]: Agreed. The two code bases have diverged too much to cleanly apply PRs.

File: test/bin/init.ps1 [hunk 6]
```diff
@@ -15,13 +15,17 @@ if ($env:CI_WINDOWS -eq $true) {
...
+        Start-FileDownload 'https://github.com/activescott/lessmsi/releases/download/v1.10.0/lessmsi-v1.10.0.zip' -FileName "$env:SCOOP_HELPERS\lessmsi.zip"
         & 7z.exe x "$env:SCOOP_HELPERS\lessmsi.zip" -o"$env:SCOOP_HELPERS\lessmsi" -y
     }
     if(!(Test-Path "$env:SCOOP_HELPERS\innounp\innounp.exe")) {
-        Start-FileDownload 'https://raw.githubusercontent.com/ScoopInstaller/Binary/master/innounp/innounp048.rar' -FileName "$env:SCOOP_HELPERS\innounp.rar"
+        Start-FileDownload 'https://raw.githubusercontent.com/ScoopInstaller/Binary/master/innounp/innounp050.rar' -FileName "$env:SCOOP_HELPERS\innounp.rar"
         & 7z.exe x "$env:SCOOP_HELPERS\innounp.rar" -o"$env:SCOOP_HELPERS\innounp" -y
     }
+    if(!(Test-Path "$env:SCOOP_HELPERS\zstd\zstd.exe")) {
+        Start-FileDownload 'https://github.com/facebook/zstd/releases/download/v1.5.0/zstd-v1.5.0-win64.zip' -FileName "$env:SCOOP_HELPERS\zstd.zip"
```
Comment (rashil2000) [inline test/bin/init.ps1:26]: @niheaven Shouldn't we use a 32bit binary (which works everywhere)? This only downloads the 64bit version.

Comment (niheaven) [reply test/bin/init.ps1:26]: Okay, I'll fix it.

Comment (pratikpc) [reply test/bin/init.ps1:26]: @niheaven https://github.com/facebook/zstd/releases/download/v1.5.0/zstd-v1.5.0-win32.zip

Comment (niheaven) [reply test/bin/init.ps1:26]: Changed in https://github.com/ScoopInstaller/Scoop/commit/47ebc6f176b0db0afeb51b4ee237a20b2d8649e9

**Pull request merged**